src/main/kotlin/first_pass/node_parser/ClassDeclarationParser.kt

```diff
@@ -0,0 +1,23 @@
+package first_pass.node_parser
+
+import cadet_model.CadetClass
+import com.github.javaparser.ast.CompilationUnit
+import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration
+import com.github.javaparser.ast.type.ClassOrInterfaceType
+
+object ClassDeclarationParser : AbstractNodeParser() {
+
+    fun instantiateClass(node: ClassOrInterfaceDeclaration, parent: CadetClass?): CadetClass {
+        return CadetClass().apply {
+            this.name = node.nameAsString
+            this.fullName = node.fullyQualifiedName.get()
+            this.parent = parent
+        }
+    }
+
```

Comment:
This was my also my concern. The ClassDeclarationParser isn't the only one that's used in the 2nd pass. If it were only that, I could easily play around it. But the problem is, 2nd pass also works with nodes because it's a visitor, and with the current algorithm/architecture, during the Resolver visit, whenever the app bumps into a MethodCallExpression, FieldAccess, or some other node which needs resolving, I have to access those Node-parsing mechanisms in order to figure out what the current node represents (name, arguments, etc). So the problem goes much deeper and this current solution is definitely up for refactoring, but I left it this way simply because it got the job done.

Reply:
Singletons have been repackaged, and are only being used by visitors from their own package.